Brokerages/Zerodha/ZerodhaBrokerage.cs

```diff
@@ -438,85 +413,92 @@ public override void Disconnect()
                 }
+                catch (Exception ex)
+                {
 
-                // Generate submitted event
-                OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, orderFee, "Zerodha Order Event") { Status = OrderStatus.Submitted });
-                Log.Trace($"Order submitted successfully - OrderId: {order.Id}");
+                    var errorMessage = $"Order failed, Order Id: {order.Id} timestamp: {order.Time} quantity: {order.Quantity} content: {ex.Message}";
+                    OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, orderFee, "Zerodha Order Event") { Status = OrderStatus.Invalid });
+                    OnMessage(new BrokerageMessageEvent(BrokerageMessageType.Warning, -1, errorMessage));
 
-                UnlockStream();
-                return true;
-            }
+                    submitted = false;
+                }
 
-            var message = $"Order failed, Order Id: {order.Id} timestamp: {order.Time} quantity: {order.Quantity} content: {orderResponse["status_message"]}";
-            OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, orderFee, "Zerodha Order Event") { Status = OrderStatus.Invalid });
-            OnMessage(new BrokerageMessageEvent(BrokerageMessageType.Warning, -1, message));
 
-            UnlockStream();
-            return false;
+                if ((string)orderResponse["status"] == "success")
+                {
+                    if (string.IsNullOrEmpty((string)orderResponse["data"]["order_id"]))
+                    {
+                        var errorMessage = $"Error parsing response from place order: {(string)orderResponse["status_message"]}";
+                        OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, orderFee, "Zerodha Order Event") { Status = OrderStatus.Invalid, Message = errorMessage });
+                        OnMessage(new BrokerageMessageEvent(BrokerageMessageType.Warning, (string)orderResponse["status_message"], errorMessage));
 
+                        submitted = false;
+                    }
+
+                    var brokerId = (string)orderResponse["data"]["order_id"];
+                    if (CachedOrderIDs.ContainsKey(order.Id))
+                    {
+                        CachedOrderIDs[order.Id].BrokerId.Clear();
+                        CachedOrderIDs[order.Id].BrokerId.Add(brokerId);
+                    }
+                    else
+                    {
+                        order.BrokerId.Add(brokerId);
+                        CachedOrderIDs.TryAdd(order.Id, order);
+                    }
+
+                    // Generate submitted event
+                    OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, orderFee, "Zerodha Order Event") { Status = OrderStatus.Submitted });
+                    Log.Trace($"Order submitted successfully - OrderId: {order.Id}");
+
+                    submitted = true;
```

Comment:
I think there is no need to set `submitted` to false, it's the default value? only need to set it on true/success, should directly return here? If it doesn't return if will emit the failure bellow? Similar comment for the other methods 👍